lib/Dialect/FIRRTL/Transforms/LowerSignatures.cpp

```diff
@@ -0,0 +1,419 @@
+// information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAttributes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/NLATable.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Dialect/HW/HWOpInterfaces.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/ADT/APSInt.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Parallel.h"
+
+#define DEBUG_TYPE "firrtl-lower-signatures"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Module Type Lowering
+//===----------------------------------------------------------------------===//
+namespace {
+
+struct AttrCache {
+  AttrCache(MLIRContext *context) {
+    nameAttr = StringAttr::get(context, "name");
+    sPortDirections = StringAttr::get(context, "portDirections");
+    sPortNames = StringAttr::get(context, "portNames");
+    sPortTypes = StringAttr::get(context, "portTypes");
+    sPortLocations = StringAttr::get(context, "portLocations");
+    sPortAnnotations = StringAttr::get(context, "portAnnotations");
+  }
+  AttrCache(const AttrCache &) = default;
+
+  StringAttr nameAttr, sPortDirections, sPortNames, sPortTypes, sPortLocations,
+      sPortAnnotations;
+};
+
+struct FieldMapEntry : public PortInfo {
+  size_t portID;
+  size_t resultID;
+  size_t fieldID;
+};
+
+using PortConversion = SmallVector<FieldMapEntry>;
+} // namespace
+
+static hw::InnerSymAttr symbolsForFieldIDRange(hw::InnerSymAttr syms,
+                                               uint64_t low, uint64_t high) {
+  // No symbols, early exit
+  if (!syms || syms.empty())
+    return {};
+
+  SmallVector<hw::InnerSymPropertiesAttr, 4> newSyms;
+  for (auto sym : syms) {
+    auto sfield = sym.getFieldID();
+    if (sfield >= low && sfield <= high) {
+      newSyms.push_back(hw::InnerSymPropertiesAttr::get(
+          syms.getContext(), sym.getName(), sfield - low,
+          sym.getSymVisibility()));
+    }
+  }
+
+  if (newSyms.empty())
+    return {};
+
+  return hw::InnerSymAttr::get(syms.getContext(), newSyms);
+}
+
+static AnnotationSet annosForFieldIDRange(const AnnotationSet &annos,
+                                          uint64_t low, uint64_t high) {
+  AnnotationSet newAnnos(annos.getContext());
+  for (auto anno : annos) {
+    auto afield = anno.getFieldID();
+    if (afield >= low && afield <= high) {
+      newAnnos.addAnnotations(anno);
+    }
```

Comment:
Yes, thanks for filing that issue!

The `partitionSymbols` method in `LowerTypes` was written the way it was to avoid that specific scaling issue ( https://github.com/llvm/circt/pull/5610#discussion_r1268282766 ).

A similar approach should work for Annotations, IIRC, and hopefully here (basically walk these once computing all you need per "lowering", which at least in LowerTypes is per-level of the type).

Reply:
And I haven't implemented the perf improvements yet.

Reply:
with better lookup tables, I've saved 5%.